File: chat-uikit/src/main/java/io/agora/chat/uikit/activities/EaseShowVideoActivity.java
package io.agora.chat.uikit.activities;

import android.net.Uri;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.Toast;


import java.io.File;

import io.agora.CallBack;
import io.agora.Error;
import io.agora.chat.ChatClient;
import io.agora.chat.ChatMessage;
import io.agora.chat.VideoMessageBody;
import io.agora.chat.uikit.R;
import io.agora.chat.uikit.base.EaseBaseActivity;
import io.agora.chat.uikit.utils.EaseFileUtils;
import io.agora.util.EMLog;

/**
 * show the video
 * 
 */
public class EaseShowVideoActivity extends EaseBaseActivity {
	private static final String TAG = "ShowVideoActivity";
	
	private RelativeLayout loadingLayout;
	private ProgressBar progressBar;
	private Uri localFilePath;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		requestWindowFeature(Window.FEATURE_NO_TITLE);
		getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
				WindowManager.LayoutParams.FLAG_FULLSCREEN);
		setContentView(R.layout.ease_showvideo_activity);
		loadingLayout = (RelativeLayout) findViewById(R.id.loading_layout);
		progressBar = (ProgressBar) findViewById(R.id.progressBar);

		final ChatMessage message = getIntent().getParcelableExtra("msg");
		if (!(message.getBody() instanceof VideoMessageBody)) {
			Toast.makeText(EaseShowVideoActivity.this, "Unsupported message body", Toast.LENGTH_SHORT).show();
			finish();
			return;
		}
		VideoMessageBody messageBody = (VideoMessageBody)message.getBody();

		localFilePath = messageBody.getLocalUri();
		EMLog.d(TAG, "localFilePath = "+localFilePath);
		EMLog.d(TAG, "local filename = "+messageBody.getFileName());

		//Check Uri read permissions
		EaseFileUtils.takePersistableUriPermission(this, localFilePath);

		if(EaseFileUtils.isFileExistByUri(this, localFilePath)) {
		    showLocalVideo(localFilePath);
		} else {
			EMLog.d(TAG, "download remote video file");
			downloadVideo(message);
		}
	}

	@Override
	public void setActivityTheme() {

	}

	private void showLocalVideo(Uri videoUri) {
		EaseShowLocalVideoActivity.actionStart(this, videoUri.toString());
		finish();
	}

	/**
	 * download video file
	 */
	private void downloadVideo(final ChatMessage message) {
		loadingLayout.setVisibility(View.VISIBLE);
		message.setMessageStatusCallback(new CallBack() {
			@Override
			public void onSuccess() {
				runOnUiThread(new Runnable() {

					@Override
					public void run() {
						loadingLayout.setVisibility(View.GONE);
						progressBar.setProgress(0);
						showLocalVideo(((VideoMessageBody)message.getBody()).getLocalUri());
					}
				});
			}

			@Override
			public void onProgress(final int progress,String status) {
				Log.d("ease", "video progress:" + progress);
				runOnUiThread(new Runnable() {

					@Override
					public void run() {
						progressBar.setProgress(progress);
					}
				});

			}

			@Override
			public void onError(final int error, String msg) {
				EMLog.e("###", "offline file transfer error:" + msg);
				Uri localFilePath = ((VideoMessageBody) message.getBody()).getLocalUri();
				String filePath = EaseFileUtils.getFilePath(EaseShowVideoActivity.this, localFilePath);
				if(TextUtils.isEmpty(filePath)) {
				    EaseShowVideoActivity.this.getContentResolver().delete(localFilePath, null, null);
				}else {
					File file = new File(filePath);
					if (file.exists()) {
						file.delete();
					}
				}

				runOnUiThread(new Runnable() {
					@Override
					public void run() {
						if (error == Error.FILE_NOT_FOUND) {
							Toast.makeText(getApplicationContext(), R.string.ease_video_expired, Toast.LENGTH_SHORT).show();
						}
					}
				});
			}
		});
		ChatClient.getInstance().chatManager().downloadAttachment(message);
	}

	@Override
	public void onBackPressed() {
		finish();
	}
 

}

File: ease-im-kit/src/main/kotlin/io/agora/uikit/feature/chat/activities/EaseShowVideoActivity.kt
package io.agora.uikit.feature.chat.activities

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.Window
import android.widget.Toast
import io.agora.uikit.EaseIM
import io.agora.uikit.R
import io.agora.uikit.base.EaseBaseActivity
import io.agora.uikit.common.ChatClient
import io.agora.uikit.common.ChatError
import io.agora.uikit.common.ChatLog
import io.agora.uikit.common.ChatMessage
import io.agora.uikit.common.ChatVideoMessageBody
import io.agora.uikit.common.extensions.hasRoute
import io.agora.uikit.common.extensions.mainScope
import io.agora.uikit.common.impl.CallbackImpl
import io.agora.uikit.common.utils.EaseFileUtils
import io.agora.uikit.common.utils.StatusBarCompat
import io.agora.uikit.databinding.EaseShowvideoActivityBinding
import kotlinx.coroutines.launch
import java.io.File

/**
 * show the video
 *
 */
class EaseShowVideoActivity : EaseBaseActivity<EaseShowvideoActivityBinding>() {
    private var localFilePath: Uri? = null

    override fun beforeSetContentView() {
        super.beforeSetContentView()
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        StatusBarCompat.hideStatusBar(this)
    }
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val message: ChatMessage? = intent.getParcelableExtra("msg")
        if (message?.body !is ChatVideoMessageBody) {
            Toast.makeText(
                this@EaseShowVideoActivity,
                "Unsupported message body",
                Toast.LENGTH_SHORT
            ).show()
            finish()
            return
        }
        val messageBody: ChatVideoMessageBody? = message.body as? ChatVideoMessageBody
        localFilePath = messageBody?.localUri
        ChatLog.d(TAG, "localFilePath = $localFilePath")
        ChatLog.d(TAG, "local filename = ${messageBody?.fileName}")

        //Check Uri read permissions
        EaseFileUtils.takePersistableUriPermission(this, localFilePath)
        if (EaseFileUtils.isFileExistByUri(this, localFilePath)) {
            showLocalVideo(localFilePath)
        } else {
            ChatLog.d(TAG, "download remote video file")
            downloadVideo(message)
        }
    }

    override fun getViewBinding(inflater: LayoutInflater): EaseShowvideoActivityBinding? {
        return EaseShowvideoActivityBinding.inflate(inflater)
    }

    override fun setActivityTheme() {}
    private fun showLocalVideo(videoUri: Uri?) {
        EaseShowLocalVideoActivity.actionStart(this, videoUri)
        finish()
    }

    /**
     * download video file
     */
    private fun downloadVideo(message: ChatMessage) {
        binding.loadingLayout.visibility = View.VISIBLE
        message.setMessageStatusCallback(CallbackImpl(
            onSuccess = {
                mainScope().launch {
                    binding.loadingLayout.visibility = View.GONE
                    binding.progressBar.progress = 0
                    showLocalVideo((message.body as ChatVideoMessageBody).localUri)
                }
            },
            onError = { code, error ->
                ChatLog.e("###", "offline file transfer error:$message")
                val localFilePath: Uri = (message.getBody() as ChatVideoMessageBody).localUri
                val filePath: String =
                    EaseFileUtils.getFilePath(this@EaseShowVideoActivity, localFilePath)
                if (TextUtils.isEmpty(filePath)) {
                    this@EaseShowVideoActivity.getContentResolver()
                        .delete(localFilePath, null, null)
                } else {
                    val file = File(filePath)
                    if (file.exists()) {
                        file.delete()
                    }
                }
                mainScope().launch {
                    if (code == ChatError.FILE_NOT_FOUND) {
                        Toast.makeText(
                            mContext,
                            R.string.ease_video_expired,
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                }
            },
            onProgress = { progress ->
                mainScope().launch {
                    binding.progressBar.progress = progress
                }
            }
        ))
        ChatClient.getInstance().chatManager().downloadAttachment(message)
    }

    override fun onBackPressed() {
        super.onBackPressed()
        finish()
    }

    companion object {
        private const val TAG = "ShowVideoActivity"
        fun actionStart(context: Context, message: ChatMessage?) {
            val intent = Intent(context, EaseShowVideoActivity::class.java)
            intent.putExtra("msg", message)
            EaseIM.getCustomActivityRoute()?.getActivityRoute(intent.clone() as Intent)?.let {
                if (it.hasRoute()) {
                    context.startActivity(it)
                    return
                }
            }
            context.startActivity(intent)
        }
    }
}